src/Razor/src/Microsoft.CodeAnalysis.Razor.CohostingShared/DevTools/CohostGeneratedDocumentContentsEndpoint.cs

```diff
@@ -0,0 +1,84 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Composition;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.ExternalAccess.Razor.Cohost;
+using Microsoft.CodeAnalysis.Razor.Cohost;
+using Microsoft.CodeAnalysis.Razor.Protocol.DevTools;
+using Microsoft.CodeAnalysis.Razor.Remote;
+using Microsoft.CodeAnalysis.Razor.Workspaces;
+
+namespace Microsoft.VisualStudio.Razor.LanguageClient.Cohost;
+
+#pragma warning disable RS0030 // Do not use banned APIs
+[Shared]
+[CohostEndpoint("razor/generatedDocumentContents")]
+[ExportRazorStatelessLspService(typeof(CohostGeneratedDocumentContentsEndpoint))]
+[method: ImportingConstructor]
+#pragma warning restore RS0030 // Do not use banned APIs
+internal sealed class CohostGeneratedDocumentContentsEndpoint(
+    IIncompatibleProjectService incompatibleProjectService,
+    IRemoteServiceInvoker remoteServiceInvoker)
+    : AbstractCohostDocumentEndpoint<DocumentContentsRequest, DocumentContentsResponse?>(incompatibleProjectService)
+{
+    private readonly IRemoteServiceInvoker _remoteServiceInvoker = remoteServiceInvoker;
+
+    protected override bool MutatesSolutionState => false;
+    protected override bool RequiresLSPSolution => true;
+
+    protected override RazorTextDocumentIdentifier? GetRazorTextDocumentIdentifier(DocumentContentsRequest request)
+        => request.TextDocument.ToRazorTextDocumentIdentifier();
+
+    protected override async Task<DocumentContentsResponse?> HandleRequestAsync(DocumentContentsRequest request, TextDocument razorDocument, CancellationToken cancellationToken)
+    {
+        string? contents = null;
+        
+        switch (request.Kind)
+        {
+            case GeneratedDocumentKind.CSharp:
+                contents = await _remoteServiceInvoker.TryInvokeAsync<IRemoteDevToolsService, string?>(
```

Comment:
The GetCSharpDocumentTextAsync method this calls should return the DocumentContentsResponse directly, as its the only thing that knows how to set the file path property (to the path of the generated file)